indico/core/emails.py

```diff
@@ -66,15 +66,26 @@ def send_email_task(task, email, log_entry=None):
             db.session.commit()
 
 
-def _rewrite_sender(msg: EmailMessage):
-    if not config.SMTP_SENDER_FALLBACK:
-        # no fallback set, cannot rewrite. let's hope all emails go through...
-        return
-
-    from_email_raw = parser.get_mailbox(msg.from_email)[0].addr_spec  # just the addr without a name part
-    if not any(fnmatch(from_email_raw, pattern) for pattern in config.SMTP_ALLOWED_SENDERS):
-        msg.extra_headers['From'] = msg.from_email
-        msg.from_email = config.SMTP_SENDER_FALLBACK
+def get_actual_sender_address(sender_address: str, reply_address: set) -> tuple[str, set]:
+    site_title = core_settings.get('site_title')
+    if not sender_address:
+        return f'{site_title} <{config.NO_REPLY_EMAIL}>', reply_address
+    if not config.SMTP_ALLOWED_SENDERS:
+        # this may result in spoofing
+        return sender_address, reply_address
+    orig_name, orig_address = parseaddr(sender_address)
+    fallback = config.SMTP_SENDER_FALLBACK
+    if any(fnmatch(orig_address, pattern) for pattern in config.SMTP_ALLOWED_SENDERS):
+        # valid sender address for which we are authorized to send emails
+        return sender_address, reply_address
+    # rewrite sender address to the fallback, and try to keep the relevant part of the original one
+    # if we have a name we use the name (the address already goes into reply-to), otherwise we use
+    # the address which looks ugly but is (probably?) better than just using the site name
+    display_name = f'{orig_name or orig_address} (via {site_title})'
```

Comment:
I wonder if the `{orig} (via {site_title})` format string should be configurable, in case someone does not want the (english) "via" there...

I was first thinking about `indico.conf`, but since this now gets called from `make_email`, we could even use proper i18n for it. I guess "via" is not really fitting for every single language after all...

cc @tomasr8